src/attachments/attachments.service.ts

```diff
@@ -16,44 +25,183 @@ export class AttachmentsService {
     @Inject(REQUEST) private request: Request,
   ) {}
 
-  async create(createAttachmentDto: CreateAttachmentDto): Promise<Attachment> {
+  async create(
+    createAttachmentDto: CreateAttachmentObsoleteDto | CreateAttachmentDto,
+  ): Promise<Attachment> {
     const username = (this.request?.user as JWTUser).username;
+
+    const convertedDto =
+      this.convertObsoleteDtoToCurrentSchema(createAttachmentDto);
     const createdAttachment = new this.attachmentModel(
-      addCreatedByFields(createAttachmentDto, username),
+      addCreatedByFields(convertedDto, username),
     );
     return createdAttachment.save();
   }
 
   async findAll(
     filter: FilterQuery<AttachmentDocument>,
   ): Promise<Attachment[]> {
-    return this.attachmentModel.find(filter).exec();
+    const convertedFilter =
+      this.convertObsoleteWhereFilterToCurrentSchema(filter);
+
+    return await this.findAllComplete({ where: convertedFilter });
+  }
+  async findAllComplete(
+    filter: FilterQuery<AttachmentDocument>,
+  ): Promise<Attachment[]> {
+    const whereFilter: FilterQuery<AttachmentDocument> = filter.where ?? {};
+    const fieldsProjection: string[] = filter.fields ?? {};
+    const limits: QueryOptions<AttachmentDocument> = filter.limits ?? {
+      limit: 10,
+      skip: 0,
+      sort: { createdAt: "desc" },
+    };
+
+    const pipeline: PipelineStage[] = [{ $match: whereFilter }];
+    if (!isEmpty(fieldsProjection)) {
+      const projection = parsePipelineProjection(fieldsProjection);
+      pipeline.push({ $project: projection });
+    }
+
+    if (!isEmpty(limits.sort)) {
+      const sort = parsePipelineSort(limits.sort);
+      pipeline.push({ $sort: sort });
+    }
+
+    pipeline.push({ $limit: limits.limit || 10 });
+
+    pipeline.push({ $skip: limits.skip || 0 });
+
+    const data = await this.attachmentModel
+      .aggregate<Attachment>(pipeline)
+      .exec();
+
+    return data;
   }
 
   async findOne(
     filter: FilterQuery<AttachmentDocument>,
     projection?: Record<string, unknown>,
   ): Promise<Attachment | null> {
-    return this.attachmentModel.findOne(filter, projection ?? {}).exec();
+    const convertedFilter =
+      this.convertObsoleteWhereFilterToCurrentSchema(filter);
+    return this.attachmentModel
+      .findOne(convertedFilter, projection ?? {})
+      .exec();
   }
 
   async findOneAndUpdate(
     filter: FilterQuery<AttachmentDocument>,
-    updateAttachmentDto: PartialUpdateAttachmentDto,
+    updateAttachmentDto:
+      | PartialUpdateAttachmentObsoleteDto
+      | PartialUpdateAttachmentDto,
```

Comment:
I would use just the current DTO here and transfer the conversion in the controller

Reply:
Then we need to force users who use the old endpoint `// PATCH /datasets/:id/attachments/:fk` to provide format of non-obsoleteDto. That's probably not a good way of handling backward compatibility

Reply:
Correct, but API V4 is not backward compatible with obsolete DTOs.
If a user wants to use Obsolete Dtos has to use API v3